Improve save file handling

cgmb opened the pull request:
I was a little confused when I saved my capture settings and then couldn't find them again. After realizing that the file was getting filtered out because I didn't write out the extension, I figured I would just add a bit of code to automatically append the file extension if necessary.

When I started doing that, I realized that other files had the same problem. I made extension appending a generic feature in getSaveFileName and updated saving for all the other file types I could identify that required a particular extension.

Then I noticed that there was a recent list for capture files. I wanted to be sure the extension appending wouldn't screw that up, so I tried playing around with it a bit. The recent files menu turned out to have a few bugs, so I fixed them. It now updates as expected and "Clear History" works.

I tried to match your style, though it feels a little weird. I also created the branch [cgmb/fix-save-file-handling-2](https://github.com/cgmb/renderdoc/compare/950c27719bd149c2195748e6d592fd6ea1d8453c...fix-save-file-handling-2) with an implementation that feels a slightly cleaner to me, but doesn't match your style as closely. If you prefer that version, I can force-push it over this branch to update the PR. If not, then merge away... or code review, or whatever.

cgmb committed:
Set default extensions for most save files

In most of these cases, the open file dialog won't even display a file
without the proper extension, so this helps ensure the user doesn't
accidentally misplace their files. The one exception is *.rdc, which
could be found without the extension, but could not be opened.

cgmb committed:
Implement "Clear History" for Recent menus

This also fixes a couple minor related bugs in the menu.

cgmb committed:
Make "Recent Capture Settings" update upon save

cgmb commented:
Opps. It appears I should have checked clang-format before submitting. I've now corrected the issues that I introduced.

baldurk commented:
Thanks for this change, I didn't realise Qt didn't do this by default. I guess it's a platform behaviour on windows or .NET that did it in C#. One thing - I believe the behaviour on windows is to add the default extension only if one isn't given (i.e. if it's just `filename` not `filename.foo`) to allow people to override the extension if they so wish. Is that worth replicating here?

If you agree on that direction it might be nice to actually make that the default behaviour even. Since e.g. there are a few places where you can save as .csv or .bin and I'd equally want to have the extension populated there.

I'm not clear exactly what the stylistic issue you're referring to in the original description. Giving a rough eye over the two branches it seems like the biggest difference is using signals/slots instead of direct callbacks? I personally don't really like the signals/slots mechanism so I don't tend to use it for my own internal functions, but I'm not trying to ban its use or anything. If that's all it is then feel free to choose either way as you like - I'll merge whatever you decide to put on this PR as both branches look fine :smile:. One minor comment is I noticed you're using the old `SIGNAL(...) / SLOT(...)` macros which aren't needed anymore on modern Qt and C++ versions, except if there's an overloaded signal to be selected.

cgmb commented:
> I believe the behaviour on windows is to add the default extension only if one isn't given

[MSDN][1] suggests that is indeed how it works in C#. I don't think there's any real standard about handling this case on any platform, so we can do pretty much whatever we want. But, the original version of renderdoc had that behaviour, so let's keep it.

...oh. You know what? I just noticed `QFileDialog::setDefaultSuffix`. It's easy and has the behaviour you describe.

> If you agree on that direction it might be nice to actually make that the default behaviour even.

If you mean using the filter string to get the full behaviour from C#, then it seems you could subclass `QFileDialog` and connect to `filterSelected` to update the `defaultSuffix`. Something similar seems to happen already when you don't set the default suffix. If your name has a suffix and you switch filters, the name will be updated to match the filter. However, I'm not thrilled about the prospect of parsing the filter string to further improve that behaviour. I'd like to call that out of scope for this change.

> I'm not clear exactly what the stylistic issue you're referring to in the original description.

The circular references between MainWindow and the other windows are pretty much what signals and slots are designed to avoid.

Renderdoc also leans heavily on `connectSlotsByName`. One example is `on_action_Exit_triggered`, which requires 5 lines of code to declare and define, but merely replaces `connect(ui->action_Exit, &QAction::triggered, this, &QWidget::close);`. I'm much more used to seeing slots as normal methods that can be connected to signals if they need to be triggered by some event. Am I right in guessing that connecting based on a naming convention is more similar to the C# events model?

But, you're getting tonnes of stuff done. Clearly it doesn't matter that much.

> One minor comment is I noticed you're using the old SIGNAL(...) / SLOT(...) macros which aren't needed anymore on modern Qt and C++ versions, except if there's an overloaded signal to be selected.

To use the new connect call, I would need to do a cast from `QWidget*` to `CaptureDialog*`. The old method works without a cast. I could add a `qobject_cast` and switch to the new style, but I don't think I want to go with that approach anyways. Having `CaptureDialog` emit a signal is cleaner, but I see now that by accessing `getCaptureDialog` in `MainWindow`'s constructor, I'm changing the order of construction and I feel a little uncomfortable with that.

I've fixed a couple mistakes I made in choosing extensions, updated `getSaveFileName` to use `setDefaultSuffix`, and settled on a hybrid of the two styles I mentioned.

[1]: https://msdn.microsoft.com/en-us/library/system.windows.forms.filedialog.defaultext(v=vs.110).aspx

baldurk commented:
> ...oh. You know what? I just noticed `QFileDialog::setDefaultSuffix`. It's easy and has the behaviour you describe.

If this would work then great, makes sense that Qt already implements it. Would this be able to handle multiple save file types e.g. the texture save dialog that will have several filters for each image type? It sounds like it probably would - set the default suffix to the initially selected filter, and then you say that changing the filter appends the suffix. As you say you can also connect `filterSelected` to update the default suffix for completeness.

> However, I'm not thrilled about the prospect of parsing the filter string to further improve that behaviour. I'd like to call that out of scope for this change.

I was thinking that since we already have our own `getSaveFileName` we can just change the interface. Instead of taking a single opaque concatenated filter string, we could take a list of string-pairs, one with the description and one with the extension. Then `getSaveFileName` can build up the filter strings and it doesn't need to do any string parsing to identify which filter is which, and get the extension from the filter.

Bear in mind though I don't think subclassing `QFileDialog` will work because e.g. on windows it won't use that code at all, it invokes the native dialog. Presumably also on OSX too if that ever becomes a thing :grin:.

It's up to you though! don't feel obliged to add that behaviour in the PR it can always be done as a follow-up, I just mentioned it as a useful further improvement.

> The circular references between MainWindow and the other windows are pretty much what signals and slots are designed to avoid.

I don't see the back references as a problem really, when the `MainWindow` needs to do something with the `CaptureDialog` it calls one way, and when the `CaptureDialog` needs to do something then it calls the other. I don't think it's turning into spaghetti, and this is a relatively uncommon circle. If necessary the other cases like `LiveCapture` handling could be broken apart so that the `MainWindow` and `LiveCapture` both have some kind of manager they both deal with, but I don't think it's a pressing concern.

That said in this case I think it should be easy enough to break the dependency entirely. The only reason you're needing to call `Populate...` from the capture dialog is because the `MainWindow` is basically 'caching' the recent files in its menu items. It's not expensive to just re-populate the list of menu items every time in a handler when the menu is opened. That way the `MainWindow` isn't caching the list of recent files and fetches it freshly from the `PersistantConfig` each time. That should solve the issue I think?

To me the signals/slots excel at 'generic' connections - when you have something that might need to be handled by different code depending on context, or multiple things might want to listen. The obvious example being any kind of widget or utility class event.

Bear in mind one other design consideration (not a hard rule but something to have in mind): if there is some function that it makes sense to expose publicly on one of the panels/windows for access elsewhere, whether it's via a slot or anything else, that heavily suggests that you probably want it exposed to python too for custom scripting - and that means exposing it via `IMainWindow` etc anyway whether or not it's a slot too.

> Renderdoc also leans heavily on connectSlotsByName. One example is on_action_Exit_triggered, which requires 5 lines of code to declare and define, but merely replaces connect(ui->action_Exit, &QAction::triggered, this, &QWidget::close);. I'm much more used to seeing slots as normal methods that can be connected to signals if they need to be triggered by some event. Am I right in guessing that connecting based on a naming convention is more similar to the C# events model?

The name-based connections are unique to Qt, there isn't anything like that in C#. Event handling is one place where I find Qt is particularly bad, but at least when comparing to .NET some of that is due to C++/moc limitations where .NET has some built-in C# features for events.

It's partly just because when you're using Qt creator, those are the names of the functions that it creates. I presume to reduce the amount of code modification Qt creator needs to do (it doesn't know necessarily where it's safe to add `QObject::connect` calls). So at least when I was initially porting most of the stuff it naturally came out like that.

Other than that though, I like being able to see where the handlers for a particular button or event are. It makes it more clear and gives you something to breakpoint on for example. Having a few more lines of code isn't that big a cost. In my experience at least there often isn't a more descriptive name than "when this event happens to this object", and a lot of the time the function isn't meaningful externally or able to be shared - so making it a named slot is superfluous.

> [...] I see now that by accessing getCaptureDialog in MainWindow's constructor, I'm changing the order of construction and I feel a little uncomfortable with that.

Yes this isn't ideal - I think it's probably safe but I wouldn't count on it, and it then might create an invisible capture dialog that isn't docked anywhere. You'd also have to handle the capture dialog being closed and deleted and then re-connecting the signal when it's opened again.

> But, you're getting tonnes of stuff done. Clearly it doesn't matter that much.

Doesn't mean I'm doing it right though :smile:. To be clear I'm not trying to say that it's My Way Or The Highway:tm: on these things, I'm not going to reject PRs because they use a slot or something. Just trying to explain how/why I'm doing things the way I am.

cgmb commented:
> Yes this isn't ideal - I think it's probably safe but I wouldn't count on it, and it then might create an invisible capture dialog that isn't docked anywhere. You'd also have to handle the capture dialog being closed and deleted and then re-connecting the signal when it's opened again.

I definitely missed the capture dialog's deletion and recreation. I had assumed it was just hidden on close. Going with the first option was definitely the right choice, then.

> Would this be able to handle multiple save file types e.g. the texture save dialog that will have several filters for each image type? It sounds like it probably would - set the default suffix to the initially selected filter, and then you say that changing the filter appends the suffix. As you say you can also connect filterSelected to update the default suffix for completeness.

The default suffix is added only once you hit Save, so sadly it won't work without some additional effort.

> Bear in mind though I don't think subclassing QFileDialog will work because e.g. on windows it won't use that code at all, it invokes the native dialog. Presumably also on OSX too if that ever becomes a thing :grin:.

That makes sense. I'm rather relying on you to test my changes on Windows. I only have Linux and OSX working on my development laptop.

> It's up to you though! don't feel obliged to add that behaviour in the PR it can always be done as a follow-up, I just mentioned it as a useful further improvement.

I'm afraid I'm out of time. I won't be able to get to this for quite a while, but I'll keep that in mind for when I can. Of course, given renderdoc's pace of development, it will probably already be done by then. :p

If there are no blocking issues, please merge.

cgmb commented:
By the way, I really appreciate the Linux support. Aside from the time commitment, one of the reasons I didn't end up contributing to Renderoc's Linux development was that I didn't actually know how to use it. It's a lot easier to contribute now that Renderdoc is a useful tool for my work.

baldurk commented:
Thanks, I don't want linux to be a second-class citizen but I still end up doing most of my development on windows just because it's still a better development environment, mainly in terms of having a good debugger available.

If you do run into any linux issues feel free to just file issues and I'm happy to tackle them, PR'ing fixes is a (welcome!) optional extra :smile:.
